stan/math/prim/fun/lub_constrain.hpp

```diff
@@ -97,27 +87,20 @@ inline return_type_t<T, L, U> lub_constrain(const T& x, const L& lb,
  *   the free scalar.
  * @throw std::domain_error if ub <= lb
  */
-template <typename T, typename L, typename U, typename S>
-inline return_type_t<T, L, U> lub_constrain(const T& x, const L& lb,
-                                            const U& ub, S& lp) {
-  using std::exp;
-  using std::log;
-  check_less("lub_constrain", "lb", lb, ub);
-  if (lb == NEGATIVE_INFTY) {
-    return ub_constrain(x, ub, lp);
-  }
-  if (ub == INFTY) {
-    return lb_constrain(x, lb, lp);
-  }
-  T inv_logit_x = inv_logit(x);
-  if (x > 0) {
-    T exp_minus_x = exp(-x);
-    lp += log(ub - lb) - x - 2 * log1p(exp_minus_x);
-  } else {
-    T exp_x = exp(x);
-    lp += log(ub - lb) + x - 2 * log1p(exp_x);
-  }
-  return fma(ub - lb, inv_logit_x, lb);
+template <typename T, typename L, typename U>
+inline auto lub_constrain(const T& x, const L& lb, const U& ub,
+                          return_type_t<T, L, U>& lp) {
+  const auto& x_ref = to_ref(x);
+  const auto& lb_ref = to_ref(lb);
+  const auto& ub_ref = to_ref(ub);
+
+  check_less("lub_constrain", "lb", value_of(lb_ref), value_of(ub_ref));
+  check_finite("lub_constrain", "lb", value_of(lb_ref));
+  check_finite("lub_constrain", "ub", value_of(ub_ref));
+  lp += sum(add(log(subtract(ub_ref, lb_ref)),
+                subtract(x_ref, multiply(2, log1p_exp(x_ref)))));
+  return eval(
+      add(elt_multiply(subtract(ub_ref, lb_ref), inv_logit(x_ref)), lb_ref));
```

Comment:
[optional] We could pull out `subtract(ub_ref, lb_ref)` here so we only do it once